Usage of non-portable strndup in cstest

### Work environment

<!-- Filling this table is mandatory -->

| Questions                                | Answers
|------------------------------------------|--------------------
| OS/arch/bits                             | Mac OS X 10.5
| Architecture                             | powerpc
| Source of Capstone                       | Source with `-DCAPSTONE_BUILD_CSTEST=ON`
| Version/git commit                       | next, 5bd05e34245eeca7833d30b677131cab6ea7a6fa
<!-- OTHER BUGS -->

### Expected behavior

cstest (`-DCAPSTONE_BUILD_CSTEST=ON`) compiles

### Actual behavior
```
[98/175] Building C object suite/cstest/CMakeFiles/libcstest.dir/src/helper.c.o
FAILED: suite/cstest/CMakeFiles/libcstest.dir/src/helper.c.o
/opt/local/bin/gcc-mp-7 -DCAPSTONE_AARCH64_SUPPORT -DCAPSTONE_ALPHA_SUPPORT -DCAPSTONE_ARM_SUPPORT -DCAPSTONE_BPF_SUPPORT -DCAPSTONE_DEBUG -DCAPSTONE_EVM_SUPPORT -DCAPSTONE_HAS_AARCH64 -DCAPSTONE_HAS_ALPHA -DCAPSTONE_HAS_ARM -DCAPSTONE_HAS_BPF -DCAPSTONE_HAS_EVM -DCAPSTONE_HAS_HPPA -DCAPSTONE_HAS_LOONGARCH -DCAPSTONE_HAS_M680X -DCAPSTONE_HAS_M68K -DCAPSTONE_HAS_MIPS -DCAPSTONE_HAS_MOS65XX -DCAPSTONE_HAS_POWERPC -DCAPSTONE_HAS_RISCV -DCAPSTONE_HAS_SH -DCAPSTONE_HAS_SPARC -DCAPSTONE_HAS_SYSTEMZ -DCAPSTONE_HAS_TMS320C64X -DCAPSTONE_HAS_TRICORE -DCAPSTONE_HAS_WASM -DCAPSTONE_HAS_X86 -DCAPSTONE_HAS_XCORE -DCAPSTONE_HAS_XTENSA -DCAPSTONE_HPPA_SUPPORT -DCAPSTONE_LOONGARCH_SUPPORT -DCAPSTONE_M680X_SUPPORT -DCAPSTONE_M68K_SUPPORT -DCAPSTONE_MIPS_SUPPORT -DCAPSTONE_MOS65XX_SUPPORT -DCAPSTONE_PPC_SUPPORT -DCAPSTONE_RISCV_SUPPORT -DCAPSTONE_SH_SUPPORT -DCAPSTONE_SPARC_SUPPORT -DCAPSTONE_SYSTEMZ_SUPPORT -DCAPSTONE_TMS320C64X_SUPPORT -DCAPSTONE_TRICORE_SUPPORT -DCAPSTONE_USE_SYS_DYN_MEM -DCAPSTONE_WASM_SUPPORT -DCAPSTONE_X86_SUPPORT -DCAPSTONE_XCORE_SUPPORT -DCAPSTONE_XTENSA_SUPPORT -I"/Users/florian/dev/capstone/include>" -I/Users/florian/dev/capstone/suite/cstest/include -I/Users/florian/dev/capstone/build/suite/cstest/extern/src/cmocka_ext/include -I/Users/florian/dev/capstone/build/suite/cstest/extern/src/libcyaml_ext/include -I/Users/florian/dev/capstone/include -g -Werror -Wall -Warray-bounds -Wshift-negative-value -Wreturn-type -Wformat -Wmissing-braces -Wunused-function -Wunused-variable -Wparentheses -Wint-in-bool-context -Wmisleading-indentation -Wno-maybe-uninitialized -Wshadow=local -MD -MT suite/cstest/CMakeFiles/libcstest.dir/src/helper.c.o -MF suite/cstest/CMakeFiles/libcstest.dir/src/helper.c.o.d -o suite/cstest/CMakeFiles/libcstest.dir/src/helper.c.o -c /Users/florian/dev/capstone/suite/cstest/src/helper.c
/Users/florian/dev/capstone/suite/cstest/src/helper.c: In function 'replace_hex':
/Users/florian/dev/capstone/suite/cstest/src/helper.c:55:13: error: implicit declaration of function 'strndup' [-Werror=implicit-function-declaration]
   tmp_tmp = strndup(tmp, orig_found - tmp);
             ^~~~~~~
/Users/florian/dev/capstone/suite/cstest/src/helper.c:55:13: error: incompatible implicit declaration of built-in function 'strndup' [-Werror]
/Users/florian/dev/capstone/suite/cstest/src/helper.c: In function 'replace_negative':
/Users/florian/dev/capstone/suite/cstest/src/helper.c:120:13: error: incompatible implicit declaration of built-in function 'strndup' [-Werror]
   tmp_tmp = strndup(tmp, orig_found - tmp);
             ^~~~~~~
cc1: all warnings being treated as errors
[100/175] Linking C static library libcapstone.a
ninja: build stopped: subcommand failed.
```

### Steps to reproduce the behavior

Compile with cmake and cstest enabled (`-DCAPSTONE_BUILD_CSTEST=ON`) on a platform that does not have strndup.

### Additional Logs, screenshots, source code,  configuration dump, ...

There are two usages of `strndup` which is specified by POSIX, but not C standard, and thus not widely available, e.g. on Mac OS X 10.5 with gcc 7. On this platform, https://github.com/capstone-engine/capstone/pull/2508 is also necessary for cstest.

--- a/suite/cstest/include/helper.h
+++ b/suite/cstest/include/helper.h
@@ -6,6 +6,8 @@
 
 #include <stddef.h>
 
+#include "../../../utils.h"
+
 #define MAX_ASM_TXT_MEM 1024
 #define X86_16 0
 #define X86_32 1
@@ -18,9 +20,5 @@ void replace_negative(char *src, size_t src_len, size_t arch_bits);
 void norm_spaces(char *str);
 void str_to_lower(char *str);
 char *cs_strndup(const char *s, size_t n);
-
-// Provided by libcapstone (utils.c). Portable strnlen replacement
-// for platforms that lack it.
-#include "../../../utils.h"
 
 #endif /* HELPER_H */

--- a/suite/cstest/include/helper.h
+++ b/suite/cstest/include/helper.h
@@ -19,4 +19,8 @@ void norm_spaces(char *str);
 void str_to_lower(char *str);
 char *cs_strndup(const char *s, size_t n);
 
+// Provided by libcapstone (utils.c). Portable strnlen replacement
+// for platforms that lack it.
+size_t cs_strnlen(const char *str, size_t n);
+
 #endif /* HELPER_H */

--- a/suite/cstest/include/helper.h
+++ b/suite/cstest/include/helper.h
@@ -21,6 +21,6 @@ char *cs_strndup(const char *s, size_t n);
 
 // Provided by libcapstone (utils.c). Portable strnlen replacement
 // for platforms that lack it.
-size_t cs_strnlen(const char *str, size_t n);
+#include "../../../utils.h"
 
 #endif /* HELPER_H */

--- a/suite/cstest/src/helper.c
+++ b/suite/cstest/src/helper.c
@@ -17,7 +17,13 @@ char *cs_strndup(const char *s, size_t n)
 	if (!s) {
 		return NULL;
 	}
-	size_t l = strnlen(s, n);
+	/* Find length up to n bytes without using strnlen(),
+	 * which is not available on older platforms such as
+	 * Mac OS X 10.5 (Leopard). */
+	size_t l = 0;
+	while (l < n && s[l] != '\0') {
+		l++;
+	}
 	if (l == SIZE_MAX) {
 		return NULL;
 	}

--- a/suite/cstest/src/helper.c
+++ b/suite/cstest/src/helper.c
@@ -17,13 +17,7 @@ char *cs_strndup(const char *s, size_t n)
 	if (!s) {
 		return NULL;
 	}
-	/* Find length up to n bytes without using strnlen(),
-	 * which is not available on older platforms such as
-	 * Mac OS X 10.5 (Leopard). */
-	size_t l = 0;
-	while (l < n && s[l] != '\0') {
-		l++;
-	}
+	size_t l = cs_strnlen(s, n);
 	if (l == SIZE_MAX) {
 		return NULL;
 	}

--- a/utils.c
+++ b/utils.c
@@ -47,6 +47,18 @@ char *cs_strdup(const char *str)
 	return (char *)memmove(new, str, len);
 }
 
+// Portable strnlen replacement for platforms that lack it
+// (e.g. Mac OS X 10.5 Leopard).
+size_t cs_strnlen(const char *str, size_t n)
+{
+	size_t l = 0;
+
+	while (l < n && str[l] != '\0')
+		l++;
+
+	return l;
+}
+
 // we need this since Windows doesn't have snprintf()
 int cs_snprintf(char *buffer, size_t size, const char *fmt, ...)
 {

--- a/utils.c
+++ b/utils.c
@@ -51,6 +51,9 @@ char *cs_strdup(const char *str)
 // (e.g. Mac OS X 10.5 Leopard).
 size_t cs_strnlen(const char *str, size_t n)
 {
+	if (!str)
+		return 0;
+
 	size_t l = 0;
 
 	while (l < n && str[l] != '\0')

--- a/utils.h
+++ b/utils.h
@@ -26,6 +26,11 @@ unsigned int count_positive8(const unsigned char *list);
 
 char *cs_strdup(const char *str);
 
+// Portable strnlen replacement. Returns the length of @str,
+// up to a maximum of @n. Needed because strnlen() is not
+// available on all platforms (e.g. Mac OS X 10.5 Leopard).
+size_t cs_strnlen(const char *str, size_t n);
+
 #define MIN(x, y) ((x) < (y) ? (x) : (y))
 
 // we need this since Windows doesn't have snprintf()

